apiserver/facades/client/charms/charmhubrepo_test.go

```diff
@@ -4,180 +4,506 @@
+	channel := corecharm.MustParseChannel("latest/stable").Normalize()
+	origin := corecharm.Origin{
+		Platform: platform,
+		Channel:  &channel,
 	}
 
-	override, matched := cp.MatchPlatform(transport.InfoChannelMap{
-		Channel: transport.Channel{
-			Platform: transport.Platform{
+	cfg, err := refreshConfig(curl, origin)
+	c.Assert(err, jc.ErrorIsNil)
+
+	ch := channel.String()
+	instanceKey := charmhub.ExtractConfigInstanceKey(cfg)
+
+	build, _, err := cfg.Build()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(build, gc.DeepEquals, transport.RefreshRequest{
+		Actions: []transport.RefreshRequestAction{{
+			Action:      "install",
+			InstanceKey: instanceKey,
+			Name:        &curl.Name,
+			Channel:     &ch,
+			Platform: &transport.RefreshRequestPlatform{
+				OS:           "ubuntu",
+				Series:       "focal",
 				Architecture: "amd64",
 			},
-		},
+		}},
+		Context: []transport.RefreshRequestContext{},
 	})
-	c.Assert(override.Arch, jc.IsFalse)
-	c.Assert(override.Series, jc.IsFalse)
-	c.Assert(matched, jc.IsTrue)
 }
 
-func (s channelPlatformSuite) TestMatchArchAll(c *gc.C) {
-	cp := channelPlatform{
-		Platform: corecharm.MustParsePlatform("amd64"),
+func (refreshConfigSuite) TestRefreshByRevision(c *gc.C) {
+	revision := 1
+	curl := charm.MustParseURL("ch:wordpress")
+	platform := corecharm.MustParsePlatform("amd64/ubuntu/focal")
+	channel := corecharm.MustParseChannel("latest/stable").Normalize()
+	origin := corecharm.Origin{
+		Platform: platform,
+		Channel:  &channel,
+		Revision: &revision,
 	}
 
-	override, matched := cp.MatchPlatform(transport.InfoChannelMap{
-		Channel: transport.Channel{
-			Platform: transport.Platform{
-				Architecture: "all",
+	cfg, err := refreshConfig(curl, origin)
+	c.Assert(err, jc.ErrorIsNil)
+
+	instanceKey := charmhub.ExtractConfigInstanceKey(cfg)
+
+	build, _, err := cfg.Build()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(build, gc.DeepEquals, transport.RefreshRequest{
+		Actions: []transport.RefreshRequestAction{{
+			Action:      "install",
+			InstanceKey: instanceKey,
+			Name:        &curl.Name,
+			Revision:    &revision,
+			Platform: &transport.RefreshRequestPlatform{
+				OS:           "ubuntu",
+				Series:       "focal",
+				Architecture: "amd64",
 			},
-		},
+		}},
+		Context: []transport.RefreshRequestContext{},
 	})
-	c.Assert(override.Arch, jc.IsTrue)
-	c.Assert(override.Series, jc.IsFalse)
-	c.Assert(matched, jc.IsTrue)
 }
 
-func (s channelPlatformSuite) TestMatchArchAMD64Revision(c *gc.C) {
-	cp := channelPlatform{
-		Platform: corecharm.MustParsePlatform("amd64"),
+func (refreshConfigSuite) TestRefreshByID(c *gc.C) {
+	id := "aaabbbccc"
+	revision := 1
+	curl := charm.MustParseURL("ch:wordpress")
+	platform := corecharm.MustParsePlatform("amd64/ubuntu/focal")
+	channel := corecharm.MustParseChannel("latest/stable").Normalize()
+	origin := corecharm.Origin{
+		ID:       id,
+		Platform: platform,
+		Channel:  &channel,
+		Revision: &revision,
 	}
```

Comment:
test has inacurate expectations, calls to refresh will fail if both channel and revision are provided: "error": { "code": "revision-conflict", "message": "Both revision and channel are present in the request for this Charm. It should be one or the other exclusively." },